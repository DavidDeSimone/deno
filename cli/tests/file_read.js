const minPowTwo = 2;
const maxPowTwo = 1024 * 1024;

const END_MARKER = "b";
const POW_TWO_MARKER = "m";
const CHAR_MARKER = "a";

const logIfNotProperEntry = (n, len) => {
  const end = String.fromCharCode(n[len - 1]);
  if (end !== END_MARKER) {
    console.log(
      `String for file input size ${len} not properly read, invalid char at position ${n
        .length - 1}, ${END_MARKER} !== ${end}`,
    );
    return;
  }

  for (let k = 0; k < len - 1; ++k) {
    const charAtKPos = String.fromCharCode(n[k]);
    if (k > 1 && (k & (k - 1)) === 0) {
      if (charAtKPos !== POW_TWO_MARKER) {
        console.log(
          `String for file input size ${len} not properly read, invalid char at position ${k}, ${POW_TWO_MARKER} !== ${charAtKPos}`,
        );
        return;
      }
    } else if (charAtKPos !== CHAR_MARKER) {
      console.log(
        `String for file input size ${len} not properly read, invalid char at position ${k}, ${CHAR_MARKER} !== ${charAtKPos}`,
      );
      return;
    }
  }
};

const textWithBufferSize = async (file, bufferSize, maxSize) => {
  console.log(`Case bufferSize: ${bufferSize}, maxSize: ${maxSize}`);
  await Deno.seek(file.rid, 0, Deno.SeekMode.Start);

  let finalBuffer = null;
  let bytesRead = 0;
  while (bytesRead < maxSize) {
    const slice = new Uint8Array(bufferSize);
    const read = await Deno.read(file.rid, slice);
    if (read === null) {
      break;
    }

    bytesRead += read;
    if (finalBuffer === null) {
      finalBuffer = slice;
    } else {
      const len = finalBuffer.length;
      const subslice = slice.subarray(0, read);
      const buffer = new Uint8Array(len + read);
      buffer.set(finalBuffer, 0);
      buffer.set(subslice, len);
      finalBuffer = buffer;
    }
  }

  const textBuffer = finalBuffer.subarray(0, bytesRead);
  const text = decoder.decode(textBuffer);
  console.log(`READ ${bytesRead}`);
  console.log(text[bytesRead - 1]);
  logIfNotProperEntry(textBuffer, bytesRead);
  console.log(bytesRead);
  console.log(text.length);
};

const encoder = new TextEncoder();
const decoder = new TextDecoder();
for (let i = minPowTwo; i <= maxPowTwo; i *= 2) {
  const fileName = await Deno.makeTempFile();
  const file = await Deno.open(fileName, { read: true, write: true });
  let data = "";
  for (let j = 0; j < i - 1; ++j) {
    if (j > 1 && (j & (j - 1)) === 0) {
      data += POW_TWO_MARKER;
    } else {
      data += CHAR_MARKER;
    }
  }

  data += END_MARKER;

  const input = encoder.encode(data);
  await Deno.write(file.rid, input);
  await textWithBufferSize(file, i, i);
  await textWithBufferSize(file, 2 * i, i);
  await textWithBufferSize(file, 300, i);
  await textWithBufferSize(file, 1024 * 16, i);
  await textWithBufferSize(file, (1024 * 16) - 1, i);
  await textWithBufferSize(file, 1024 * 32, i);

  await Deno.remove(fileName);
  await Deno.close(file.rid);
}
