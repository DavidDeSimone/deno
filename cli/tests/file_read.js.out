Case bufferSize: 2, maxSize: 2
READ 2
b
2
2
Case bufferSize: 4, maxSize: 2
READ 2
b
2
2
Case bufferSize: 300, maxSize: 2
READ 2
b
2
2
Case bufferSize: 16384, maxSize: 2
READ 2
b
2
2
Case bufferSize: 16383, maxSize: 2
READ 2
b
2
2
Case bufferSize: 32768, maxSize: 2
READ 2
b
2
2
Case bufferSize: 4, maxSize: 4
READ 4
b
4
4
Case bufferSize: 8, maxSize: 4
READ 4
b
4
4
Case bufferSize: 300, maxSize: 4
READ 4
b
4
4
Case bufferSize: 16384, maxSize: 4
READ 4
b
4
4
Case bufferSize: 16383, maxSize: 4
READ 4
b
4
4
Case bufferSize: 32768, maxSize: 4
READ 4
b
4
4
Case bufferSize: 8, maxSize: 8
READ 8
b
8
8
Case bufferSize: 16, maxSize: 8
READ 8
b
8
8
Case bufferSize: 300, maxSize: 8
READ 8
b
8
8
Case bufferSize: 16384, maxSize: 8
READ 8
b
8
8
Case bufferSize: 16383, maxSize: 8
READ 8
b
8
8
Case bufferSize: 32768, maxSize: 8
READ 8
b
8
8
Case bufferSize: 16, maxSize: 16
READ 16
b
16
16
Case bufferSize: 32, maxSize: 16
READ 16
b
16
16
Case bufferSize: 300, maxSize: 16
READ 16
b
16
16
Case bufferSize: 16384, maxSize: 16
READ 16
b
16
16
Case bufferSize: 16383, maxSize: 16
READ 16
b
16
16
Case bufferSize: 32768, maxSize: 16
READ 16
b
16
16
Case bufferSize: 32, maxSize: 32
READ 32
b
32
32
Case bufferSize: 64, maxSize: 32
READ 32
b
32
32
Case bufferSize: 300, maxSize: 32
READ 32
b
32
32
Case bufferSize: 16384, maxSize: 32
READ 32
b
32
32
Case bufferSize: 16383, maxSize: 32
READ 32
b
32
32
Case bufferSize: 32768, maxSize: 32
READ 32
b
32
32
Case bufferSize: 64, maxSize: 64
READ 64
b
64
64
Case bufferSize: 128, maxSize: 64
READ 64
b
64
64
Case bufferSize: 300, maxSize: 64
READ 64
b
64
64
Case bufferSize: 16384, maxSize: 64
READ 64
b
64
64
Case bufferSize: 16383, maxSize: 64
READ 64
b
64
64
Case bufferSize: 32768, maxSize: 64
READ 64
b
64
64
Case bufferSize: 128, maxSize: 128
READ 128
b
128
128
Case bufferSize: 256, maxSize: 128
READ 128
b
128
128
Case bufferSize: 300, maxSize: 128
READ 128
b
128
128
Case bufferSize: 16384, maxSize: 128
READ 128
b
128
128
Case bufferSize: 16383, maxSize: 128
READ 128
b
128
128
Case bufferSize: 32768, maxSize: 128
READ 128
b
128
128
Case bufferSize: 256, maxSize: 256
READ 256
b
256
256
Case bufferSize: 512, maxSize: 256
READ 256
b
256
256
Case bufferSize: 300, maxSize: 256
READ 256
b
256
256
Case bufferSize: 16384, maxSize: 256
READ 256
b
256
256
Case bufferSize: 16383, maxSize: 256
READ 256
b
256
256
Case bufferSize: 32768, maxSize: 256
READ 256
b
256
256
Case bufferSize: 512, maxSize: 512
READ 512
b
512
512
Case bufferSize: 1024, maxSize: 512
READ 512
b
512
512
Case bufferSize: 300, maxSize: 512
READ 512
b
512
512
Case bufferSize: 16384, maxSize: 512
READ 512
b
512
512
Case bufferSize: 16383, maxSize: 512
READ 512
b
512
512
Case bufferSize: 32768, maxSize: 512
READ 512
b
512
512
Case bufferSize: 1024, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 2048, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 300, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 16384, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 16383, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 32768, maxSize: 1024
READ 1024
b
1024
1024
Case bufferSize: 2048, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 4096, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 300, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 16384, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 16383, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 32768, maxSize: 2048
READ 2048
b
2048
2048
Case bufferSize: 4096, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 8192, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 300, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 16384, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 16383, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 32768, maxSize: 4096
READ 4096
b
4096
4096
Case bufferSize: 8192, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 16384, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 300, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 16384, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 16383, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 32768, maxSize: 8192
READ 8192
b
8192
8192
Case bufferSize: 16384, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 32768, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 300, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 16384, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 16383, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 32768, maxSize: 16384
READ 16384
b
16384
16384
Case bufferSize: 32768, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 65536, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 300, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 16384, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 16383, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 32768, maxSize: 32768
READ 32768
b
32768
32768
Case bufferSize: 65536, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 131072, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 300, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 16384, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 16383, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 32768, maxSize: 65536
READ 65536
b
65536
65536
Case bufferSize: 131072, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 262144, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 300, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 16384, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 16383, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 32768, maxSize: 131072
READ 131072
b
131072
131072
Case bufferSize: 262144, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 524288, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 300, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 16384, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 16383, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 32768, maxSize: 262144
READ 262144
b
262144
262144
Case bufferSize: 524288, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 1048576, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 300, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 16384, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 16383, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 32768, maxSize: 524288
READ 524288
b
524288
524288
Case bufferSize: 1048576, maxSize: 1048576
READ 1048576
b
1048576
1048576
Case bufferSize: 2097152, maxSize: 1048576
READ 1048576
b
1048576
1048576
Case bufferSize: 300, maxSize: 1048576
READ 1048576
b
1048576
1048576
Case bufferSize: 16384, maxSize: 1048576
READ 1048576
b
1048576
1048576
Case bufferSize: 16383, maxSize: 1048576
READ 1048576
b
1048576
1048576
Case bufferSize: 32768, maxSize: 1048576
READ 1048576
b
1048576
1048576
